File: dna/dna.go
package main

import (
	"bufio"
	"fmt"
	"io"
	"os"
	"strconv"
	"strings"
	"time"
)

type dna struct {
	start int32
	end   int32
	d     string
}

type vertex struct {
	children    map[rune]int
	leaf        bool
	parent      int
	parentChar  rune
	suffixLink  int
	endWordLink int
	idxs        []int
	pattern     string
	visited 		bool
}

type ahoTrie struct {
	trie []*vertex
	size int
	root int
}

func (aho *ahoTrie) Init() {
	aho.trie = append(aho.trie, &vertex{
		children: make(map[rune]int),
	})
	aho.size++
	aho.root = 0
}

func (aho *ahoTrie) add(pattern string, idx int) {
	currentvertex := aho.root

	for _, v := range pattern {
		if _, ok := aho.trie[currentvertex].children[v]; !ok {
			aho.trie = append(aho.trie, &vertex{
				suffixLink: -1,
				endWordLink: -1,
				parent:     currentvertex,
				parentChar: v,
				idxs:       []int{},
				children:   make(map[rune]int),
				visited: false,
			})

			aho.trie[currentvertex].children[v] = aho.size
			aho.size++
		}

		currentvertex = aho.trie[currentvertex].children[v]
	}

	aho.trie[currentvertex].leaf = true
	aho.trie[currentvertex].pattern = pattern
	aho.trie[currentvertex].idxs = append(aho.trie[currentvertex].idxs, idx)
}

func (aho *ahoTrie) calculateSuffixLink(vertex int) {
	if vertex == aho.root || aho.trie[vertex].parent == aho.root {
		aho.trie[vertex].suffixLink = aho.root
		aho.trie[vertex].endWordLink = aho.root

		if vertex != aho.root && aho.trie[vertex].leaf {
			aho.trie[vertex].endWordLink = vertex
		}

		return
	}

	parentSuffix := aho.trie[aho.trie[vertex].parent].suffixLink

	for {
		suffixNode := aho.trie[parentSuffix]

		if _, ok := suffixNode.children[aho.trie[vertex].parentChar]; ok {
			aho.trie[vertex].suffixLink = suffixNode.children[aho.trie[vertex].parentChar]
			break
		}

		if parentSuffix == aho.root {
			aho.trie[vertex].suffixLink = aho.root
			break
		}

		parentSuffix = suffixNode.suffixLink
	}

	if aho.trie[vertex].leaf {
		aho.trie[vertex].endWordLink = vertex
	} else {
		aho.trie[vertex].endWordLink = aho.trie[aho.trie[vertex].suffixLink].endWordLink
	}
}

type queue struct {
	values []int
	count  int
}

func (q *queue) Init(value int) {
	q.values = make([]int, 0)

	q.values = append(q.values, value)

	q.count = 1
}

// remove old element
func (q *queue) Dequeue() int {
	last := q.values[0]
	q.values = q.values[1:]

	q.count--

	return last
}

func (q *queue) Enqueue(value int) {
	q.values = append(q.values, value)
	q.count++
}

func (aho *ahoTrie) Prepare() {
	vertexQueue := queue{}
	vertexQueue.Init(aho.root)

	for vertexQueue.count > 0 {
		currentvertex := vertexQueue.Dequeue()

		aho.calculateSuffixLink(currentvertex)

		for key := range aho.trie[currentvertex].children {
			curr := aho.trie[currentvertex].children[key]
			if !aho.trie[curr].visited {
				aho.trie[curr].visited = true
				vertexQueue.Enqueue(curr)
			}
		}
	}
}

func (aho *ahoTrie) ProcessString(text string, first int, last int) []int {
	curentState := aho.root

	result := []int{}

	for j := 0; j < len(text); j++ {
		for {
			if _, ok := aho.trie[curentState].children[rune(text[j])]; ok {
				curentState = aho.trie[curentState].children[rune(text[j])]
				break
			}

			if curentState == aho.root {
				break
			}

			curentState = aho.trie[curentState].suffixLink
		}
		checkstate := curentState

		for {
			checkstate = aho.trie[checkstate].endWordLink

			if checkstate == aho.root {
				break
			}

			idxs := aho.trie[checkstate].idxs

			if len(idxs) == 1 {
				idx := idxs[0]

				if idx >= first && idx <= last {
					result = append(result, idx)
				}
			} else {
				for _, idx := range idxs {
					if idx > last {
						break
					}

					if idx >= first {
						result = append(result, idx)
					}
				}
			}

			checkstate = aho.trie[checkstate].suffixLink
		}
	}

	return result
}

func dnaAnalysis(genes []string, health []int32, dnas []*dna) {
	aho := ahoTrie{}
	aho.Init()

	start := time.Now()
	for i, v := range genes {
		aho.add(v, i)
	}

	aho.Prepare()
	elapsed := time.Since(start)
	fmt.Printf("Prepare took %s\n", elapsed)

	min := int(^uint(0) >> 1)
	max := 0

	for _, seq := range dnas {
		matches := aho.ProcessString(seq.d, int(seq.start), int(seq.end))

		total := 0

		for _, idx := range matches {
			total += int(health[idx])
		}

		if total > max {
			max = total
		}

		if total < min {
			min = total
		}
	}

	fmt.Printf("%d %d\n", min, max)
}

func main() {
	reader := bufio.NewReaderSize(os.Stdin, 16*1024*1024)

	nTemp, err := strconv.ParseInt(strings.TrimSpace(readLine(reader)), 10, 64)
	checkError(err)
	n := int32(nTemp)

	genesTemp := strings.Split(strings.TrimSpace(readLine(reader)), " ")

	var genes []string

	for i := 0; i < int(n); i++ {
		genesItem := genesTemp[i]
		genes = append(genes, genesItem)
	}

	healthTemp := strings.Split(strings.TrimSpace(readLine(reader)), " ")

	var health []int32

	for i := 0; i < int(n); i++ {
		healthItemTemp, err := strconv.ParseInt(healthTemp[i], 10, 64)
		checkError(err)
		healthItem := int32(healthItemTemp)
		health = append(health, healthItem)
	}

	sTemp, err := strconv.ParseInt(strings.TrimSpace(readLine(reader)), 10, 64)
	checkError(err)
	s := int32(sTemp)
	var dnas []*dna

	for sItr := 0; sItr < int(s); sItr++ {
		firstMultipleInput := strings.Split(strings.TrimSpace(readLine(reader)), " ")

		firstTemp, err := strconv.ParseInt(firstMultipleInput[0], 10, 64)
		checkError(err)
		first := int32(firstTemp)

		lastTemp, err := strconv.ParseInt(firstMultipleInput[1], 10, 64)
		checkError(err)
		last := int32(lastTemp)

		d := firstMultipleInput[2]
		input := dna{start: first, end: last, d: d}

		dnas = append(dnas, &input)
	}

	dnaAnalysis(
		genes,
		health,
		dnas,
	)
}

func readLine(reader *bufio.Reader) string {
	str, _, err := reader.ReadLine()
	if err == io.EOF {
		return ""
	}

	return strings.TrimRight(string(str), "\r\n")
}

func checkError(err error) {
	if err != nil {
		panic(err)
	}
}

File: dna/dna.py
#!/bin/python3

class Vertex:
  def __init__(self, parentIndex=None, char=None):
    self.char = char
    self.idx = []
    self.parentIndex = parentIndex
    self.child = {}
    self.suffix = None
    self.endlink = None
    self.pattern = None
    self.leaf = False

class Trie:
  def __init__(self):
    root = Vertex()

    self.trie = [root]
    self.size = 1
    self.root = 0

  def add(self, pattern, idx):
    currentVertex = self.root

    for char in pattern:
      if not char in self.trie[currentVertex].child:
        self.trie.append(Vertex(
          currentVertex,
          char
        ))

        self.trie[currentVertex].child[char] = self.size
        self.size += 1

      currentVertex = self.trie[currentVertex].child[char]
    
    self.trie[currentVertex].leaf = True
    self.trie[currentVertex].pattern = pattern
    self.trie[currentVertex].idx.append(idx)

  # find suffix and endlink
  def build_suffix_and_endlink(self):
    queue = [
      self.root
    ]

    while queue:
      s = queue.pop(0)

      # print("iterate over {}".format(self.trie[s].char))
      self.calculate_suffix(s)
      
      for key, pos in self.trie[s].child.items():
        vertex = self.trie[s].child[key]
        if vertex:
          queue.append(vertex)

  def set_root_and_children_sufix(self, vertex):
    vertex_node = self.trie[vertex]
    vertex_node.suffix = self.root
    vertex_node.endlink = self.root

    if vertex_node != self.root and vertex_node.leaf:
      vertex_node.endlink = vertex
      

  # vertex is a index
  def calculate_suffix(self, vertex):
    vertex_node = self.trie[vertex]

    # if root or his childrens just set sufix as root
    if vertex == self.root or vertex_node.parentIndex == self.root:
      self.set_root_and_children_sufix(vertex)
      return

    # backtracking
    parentSuffix = self.trie[vertex_node.parentIndex].suffix
    while True:
      sufixNode = self.trie[parentSuffix]
      # if found vertex as child of sufixNode, set them as sufix
      if self.trie[vertex].char in sufixNode.child:
        vertex_node.suffix = sufixNode.child[self.trie[vertex].char]
        break

      # if parent sufix is root, there are no matches
      if parentSuffix == self.root:
        vertex_node.suffix = self.root
        break
      
      # keep coming back, navigating by sufix
      parentSuffix = sufixNode.suffix
    
    if vertex_node.leaf:
      vertex_node.endlink = vertex
    else:
      vertex_node.endlink = self.trie[vertex_node.suffix].endlink

  def process(self, text, first, last):
    state = self.root

    result = []

    for char in text:
      while True:
        if char in self.trie[state].child:
          state = self.trie[state].child[char]
          break
      
        if state == self.root:
          break
        
        state = self.trie[state].suffix

      check_state = state

      while True:
        check_state = self.trie[check_state].endlink

        if check_state == self.root:
          break
        
        indexes = self.trie[check_state].idx
        for i in indexes:
          if i > last:
            break
          if i >= first:
            result.append(i)

        check_state = self.trie[check_state].suffix
    
    return result

class DNA:
  def __init__(self, first, last, code):
    self.first = first
    self.last = last
    self.code = code

  def __str__(self):
    return "{}-{} | {}".format(self.first, self.last, self.code)


def dna(genes, health, dnas):
  t = Trie()

  for idx, g in enumerate(genes):
    t.add(g, idx)

  t.build_suffix_and_endlink()

  minimum = float("inf")
  maximum = 0

  for dna in dnas:
    matches = t.process(dna.code, dna.first, dna.last)

    total = 0
    for s in matches:
      total += health[s]

    if total > maximum:
      maximum = total
    elif total < minimum:
      minimum = total

  print("{} {}".format(minimum, maximum))


if __name__ == '__main__':
    n = int(input().strip())

    genes = input().rstrip().split()

    health = list(map(int, input().rstrip().split()))

    dnas = []

    s = int(input().strip())

    for s_itr in range(s):
        first_multiple_input = input().rstrip().split()

        first = int(first_multiple_input[0])

        last = int(first_multiple_input[1])

        d = first_multiple_input[2]

        dnas.append(DNA(first, last, d))

    dna(genes, health, dnas)

